File: aiml_creator.cpp
#include <bits/stdc++.h>

#define OUTPUT_FILE "output.aiml"

#define debug if(true)

using namespace std;

struct variants {
	vector<string> v;
	variants(){}
	variants(string file) {
		ifstream in (file + ".set", ios::in);
		if(in.is_open() == false) throw "Error opening file";
		while(in.eof() == false) {
			string s;
			getline(in, s);
			v.push_back(s);
		}
		in.close();
	}

	string operator[](int pos) {
		if(pos > v.size()) {
			cerr << "Position " << pos << "is invalid\n";
			return "";
		}
		return v[pos];
	}

	int size() {
		return v.size();
	}
};
int flag = 0;

map<string, variants> var;

string create_pattern(string & que, string & ans) {
	string res;
	res += "\t<category>\n";
	res += "\t\t<pattern> ";
	

	res += que;
	res += "</pattern>\n\t\t<template> ";
	res += ans;
	res += "</template>\n\t</category>\n\n";

	return res;
}

void build_question(vector<string> & list, string & que, int p, string s, string & ans) {
	if(p == que.size()) {
		list.push_back(create_pattern(s, ans));
		if(flag == 0) {
			ans = "<srai> " + s + " </srai>";
			flag = 1;
		}
		return;
	}

	if(que[p] == '?') {
		string x;
		for(int i = p+1; que[i] != '?'; ++i) x += que[i];

		if(var.count(x) == 0) try {
			var[x] = variants(x);
		} catch(const char * message) {
			cout << message << endl;
			exit(0);
		}
		for(int i = 0; i < var[x].size(); ++i)
			build_question(list, que, p + x.size() + 2, s + var[x][i], ans);
	} else build_question(list, que, p+1, s + que[p], ans);
}
vector<string> processLine(string que, string ans) {
	vector<string> v;

	flag = 0;

	build_question(v, que, 0, "", ans);
	return v;
}

int main(int argc, char * argv[]) {
	string res = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<aiml>\n";
	string out_file;

	if(argc != 2 and argc != 3) {
		printf("Usage: %s <input text file> [output file] \n", argv[0]);
		return 0;
	}

	string file = string(argv[1]);
	if(argc == 3) out_file = string(argv[2]);
	else out_file = OUTPUT_FILE;

	ifstream in (file, ios::in);
	if(in.is_open()) {
		while(in.eof() == false) {
			string question, answer;
			
			while(in.eof() == false and (question.size() == 0 or question[0] == '\n')) getline(in, question);
			while(in.eof() == false and (answer.size() == 0 or answer[0] == '\n')) getline(in, answer);
			if(in.eof()) break;
			
			vector<string> ret = processLine(question, answer);
			for(string s : ret) res += s;

		}
		in.close();
		res += "</aiml>\n";
	} else {
		printf("Erro ao abrir arquivo de entrada\n");
		return 0;
	}

	ofstream out (out_file, ios::out);
	if(out.is_open()) {
		out << res;
		out.close();
	} else {
		printf("Erro ao abrir arquivo de saida\n");
		return 0;
	}
}

File: aiml_creator.py
import sys

def create_category(pattern, that, template):
	patterns = []
	# create all configurations of pattern
	cnt = 0
	for c in pattern:
		cnt += 1 if c == '^' else 0

	for i in range(1 << cnt):
		s = ""
		ind = 0
		for c in pattern:
			if c == '^':
				if ((i >> ind)&1) == 1:
					s += '*'
				ind += 1
			else: s += c

		s2 = ""
		last = ' '
		for c in s:
			if c != ' ' or c != last:
				s2 += c
			last = c
		while s2[-1] == ' ': s2 = s2[:-1]
		patterns.append(s2);

	str = ""
	for pat in patterns:
		str += "\t<category>\n"
		str += "\t\t<pattern>"+ pat +"</pattern>\n"
		if that != "": str += "\t\t<that>"+ that +"</that>\n"
		str += "\t\t<template>"+ template +"</template>\n"
		str += "\t</category>\n"
	return str

result = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<aiml>\n"
question = ""
that = ""
for line in sys.stdin:
	if line[0] == '\n': continue
	line = line.replace("\n", "");
	if line[0] == '-': # question
		if question != "":
			sys.exit("Something is wrong with your format")
		question = line[1:]
	if line[0] == '@': # question
		if that != "":
			sys.exit("Something is wrong with your format")
		that = line[1:]
	elif line[0] == '>': # answer
		if question == "":
			sys.exit("Something is wrong with your format")
		answer = line[1:]
		result += create_category(question, that, answer)
		question = ""
		that = ""

result += "</aiml>"
print(result)